<a href="https://colab.research.google.com/github/repoocsov/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [42]:
# In which issues do republicans and democrats prove to have a statistically significant difference (95% certainty)
import pandas as pd
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', header=None, na_values='?')
df.head()

0    1  2  3    4    5  6  7  8  9  10   11   12 13 14 15   16
0  republican    n  y  n    y    y  y  n  n  n  y  NaN    y  y  y  n    y
1  republican    n  y  n    y    y  y  n  n  n  n    n    y  y  y  n  NaN
2    democrat  NaN  y  y  NaN    y  y  n  n  n  n    y    n  y  y  n    n
3    democrat    n  y  y    n  NaN  y  n  n  n  n    y    n  y  n  n    y
4    democrat    y  y  y    n    y  y  n  n  n  n    y  NaN  y  y  y    y

In [43]:
# Fixing headers, converting strings into a numeric format, and seperating democrats and republicans into two seperate dataframes

# Designating headers
headers = ['party', 'handicapped-infants', 'water-project', 
'budget', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration', 
'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free', 
'south-africa']
df.columns = headers

# Converting to numeric format
df = df.replace({'n': 0, 'y': 1})
df.head()

# Creating two seperate dataframes
condition = df['party'] == 'republican'
republicans = df[condition]
republicans.head(20)

condition = df['party'] == 'democrat'
democrats = df[condition]
democrats.head(20)

# 267 dems
# 168 repubs

party  handicapped-infants  ...  duty-free  south-africa
2   democrat                  NaN  ...        0.0           0.0
3   democrat                  0.0  ...        0.0           1.0
4   democrat                  1.0  ...        1.0           1.0
5   democrat                  0.0  ...        1.0           1.0
6   democrat                  0.0  ...        1.0           1.0
9   democrat                  1.0  ...        NaN           NaN
12  democrat                  0.0  ...        NaN           NaN
13  democrat                  1.0  ...        1.0           NaN
16  democrat                  1.0  ...        0.0           1.0
17  democrat                  1.0  ...        1.0           1.0
19  democrat                  1.0  ...        1.0           1.0
20  democrat                  1.0  ...        1.0           1.0
21  democrat                  1.0  ...        1.0           1.0
22  democrat                  1.0  ...        1.0           1.0
23  democrat                  1.0  ...        1.0           1.0
24  democrat                  1.0  ...        1.0           NaN
25  democrat                  1.0  ...        1.0           1.0
26  democrat                  1.0  ...        1.0           1.0
27  democrat                  1.0  ...        1.0           1.0
29  democrat                  1.0  ...        1.0           1.0

[20 rows x 17 columns]

In [44]:
import numpy as np

# Starting with the 'crime' bill
dem_crime = pd.DataFrame(democrats['crime'])
repub_crime = pd.DataFrame(republicans['crime'])


dem_crime.mean()
# Alternatively
np.mean(dem_crime)

repub_crime.mean()
# Alternatively
np.mean(repub_crime)

# Our sample means (Can't infer back to population with this data)
# Dem "average" vote = 0.350195
# Repub "average" vote = 0.981366
# Now we run a two sample t-test to check that these means are statistically significant and not due to random chance (measure at a 95% confident level)

crime    0.981366
dtype: float64

##**Two Sample T-test for the "crime" bill**

**Null Hypothesis:** There is no difference in voting rates between democrat and republican house of representatives on the 'crime' bill. (support is equal)

$\bar{x}_{1} == \bar{x}_{2}$

Where $\bar{x}_{1}$ is the mean republican vote and $\bar{x}_{2}$ is the mean democrat vote.

**Alternative Hypothesis:** Voting rates between democrat and republican house of representatives will differ on the 'crime' bill.

$\bar{x}_{1} \neq \bar{x}_{2}$

**Confidence Level:** 99%

In [46]:
# http://www.homepage.divms.viowa.edu/~mbognar/applets/t.html
from scipy.stats import ttest_ind
ttest_ind(republicans['crime'], democrats['crime'], nan_policy='omit')

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

##**Conclusion:**

Due to observing a t-statistic of 16.34 and a resulting p-value of 9.95e-47 (less than .01), we reject the null hypothesis that the voting rates between republican and democrat house of representatives on the 'crime' bill are equal, and suggest the alternative hypothesis, that there is a statistically significant difference.


**Assignment:**
~~Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01~~

##**Two Sample T-test for the "duty-free" bill**

**Null Hypothesis:** There is no difference in voting rates between democrat and republican house of representatives on the 'duty-free' bill. (support is equal)

$\bar{x}_{1} == \bar{x}_{2}$

Where $\bar{x}_{1}$ is the mean republican vote and $\bar{x}_{2}$ is the mean democrat vote.

**Alternative Hypothesis:** Voting rates between democrat and republican house of representatives will differ on the 'duty-free' bill.

$\bar{x}_{1} \neq \bar{x}_{2}$

**Confidence Level:** 99%

In [52]:
dem_duty = pd.DataFrame(democrats['duty-free'])
repub_duty = pd.DataFrame(republicans['duty-free'])


print(dem_duty.mean()) # Dem mean vote on the duty-free bill = .637
repub_duty.mean() # Repub mean vote on the duty-free bill = .0897

duty-free    0.63745
dtype: float64


duty-free    0.089744
dtype: float64

In [53]:
ttest_ind(republicans['duty-free'], democrats['duty-free'], nan_policy='omit')

Ttest_indResult(statistic=-12.853146132542978, pvalue=5.997697174347365e-32)

##**Conclusion:**

Due to observing a t-statistic of -12.85 and a resulting p-value of 5.997e-32 (less than .01), we reject the null hypothesis that the voting rates between republican and democrat house of representatives on the 'crime' bill are equal, and suggest the alternative hypothesis, that there is a statistically significant difference.


**Assignment:**
~~Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01~~


##**Two Sample T-test for the "budget" bill**

**Null Hypothesis:** There is no difference in voting rates between democrat and republican house of representatives on the 'budget' bill. (support is equal)

$\bar{x}_{1} == \bar{x}_{2}$

Where $\bar{x}_{1}$ is the mean republican vote and $\bar{x}_{2}$ is the mean democrat vote.

**Alternative Hypothesis:** Voting rates between democrat and republican house of representatives will differ on the 'budget' bill.

$\bar{x}_{1} \neq \bar{x}_{2}$

**Confidence Level:** 99%

In [66]:
dem_budget = pd.DataFrame(democrats['immigration'])
repub_budget = pd.DataFrame(republicans['immigration'])

print(dem_budget.mean()) # Dem mean vote on the immigration bill = .471
repub_budget.mean() # Repub mean vote on the immigration bill = .557

immigration    0.471483
dtype: float64


immigration    0.557576
dtype: float64

In [67]:
ttest_ind(republicans['immigration'], democrats['immigration'], nan_policy='omit')

Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

##**Conclusion:**

Due to observing a t-statistic of 1.735 and a resulting p-value of .0833 (greater than .01), we fail to reject the null hypothesis that voting rates between republican and democrat house of representatives on the 'crime' bill are equal.


**Assignment:**
~~Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)~~